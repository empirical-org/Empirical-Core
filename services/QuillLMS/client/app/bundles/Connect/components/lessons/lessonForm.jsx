import { ContentState, EditorState } from 'draft-js';
import React from 'react';
import { connect } from 'react-redux';
import SelectSearch from 'react-select-search';
import _ from 'underscore';
import ChooseModelContainer from './chooseModelContainer.jsx';

import {
  hashToCollection,
  SortableList,
  TextEditor
} from '../../../Shared/index';

class LessonForm extends React.Component {
  constructor(props) {
    super(props)

    const { currentValues, } = props;

    this.state = {
      name: currentValues ? currentValues.name : '',
      introURL: currentValues ? currentValues.introURL || '' : '',
      landingPageHtml: currentValues ? currentValues.landingPageHtml || '' : '',
      selectedQuestions: currentValues && currentValues.questions ? currentValues.questions : [],
      flag: currentValues ? currentValues.flag : 'alpha',
      questionType: currentValues ? currentValues.questionType || 'questions' : 'questions',
      modelConceptUID: currentValues ? currentValues.modelConceptUID : null
    }
  }

  handleChange = value => {
    const currentSelectedQuestions = this.state.selectedQuestions;
    let newSelectedQuestions;
    const changedQuestion = currentSelectedQuestions.find(q => q.key === value);
    if (!changedQuestion) {
      newSelectedQuestions = currentSelectedQuestions.concat([{ key: value, questionType: this.state.questionType, }]);
    } else {
      newSelectedQuestions = _.without(currentSelectedQuestions, changedQuestion);
    }
    this.setState({ selectedQuestions: newSelectedQuestions, });
  };

  handleLPChange = e => {
    this.setState({ landingPageHtml: e, });
  };

  handleSearchChange = value => {
    this.handleChange(value);
  };

  handleSelect = e => {
    this.setState({ flag: e.target.value, });
  };

  handleSelectQuestionType = e => {
    this.setState({ questionType: e.target.value, });
  };

  handleStateChange = (key, event) => {
    const changes = {};
    changes[key] = event.target.value;
    this.setState(changes);
  };

  sortCallback = sortInfo => {
    const newOrder = sortInfo.map(item => Object.assign({key: item.key, questionType: item.props.questionType}));
    this.setState({ selectedQuestions: newOrder, });
  };

  submit = () => {
    const { name, selectedQuestions, landingPageHtml, flag, modelConceptUID, questionType } = this.state
    this.props.submit({
      name,
      questions: selectedQuestions,
      landingPageHtml,
      flag,
      modelConceptUID,
      questionType
    });
  };

  renderQuestionSelect = () => {
    let questions;
    // select changes based on whether we are looking at 'questions' (should be refactored to sentenceCombining) or sentenceFragments
    if (this.state.selectedQuestions && this.state.selectedQuestions.length) {
      const questionsList = this.state.selectedQuestions.map((question) => {
        const questionobj = this.props[question.questionType].data[question.key];
        const prompt = questionobj ? questionobj.prompt : 'Question No Longer Exists';
        const promptOrTitle = question.questionType === 'titleCards' ? questionobj.title : prompt
        return (
          <p className="sortable-list-item" id={question.key} key={question.key} questionType={question.questionType}>
            {promptOrTitle}
            {'\t\t'}
            <button onClick={this.handleChange.bind(null, question.key)}>Delete</button>
          </p>
        );
      });
      return <SortableList data={questionsList} key={this.state.selectedQuestions.length} sortCallback={this.sortCallback} />;
    } else {
      return <div>No questions</div>;
    }
  };

  renderSearchBox = () => {
    // options changes based on whether we are looking at 'questions' (should be refactored to sentenceCombining) or sentenceFragments
    const questionType = this.state.questionType;
    let options = hashToCollection(this.props[questionType].data);
    const concepts = this.props.concepts.data[0];
    let formatted
    if (options.length > 0) {
      if (questionType !== 'titleCards') {
        options = _.filter(options, option => _.find(concepts, { uid: option.conceptID, }) && (option.flag !== "archived")); // filter out questions with no valid concept
        formatted = options.map(opt => ({ name: opt.prompt.replace(/(<([^>]+)>)/ig, '').replace(/&nbsp;/ig, ''), value: opt.key, }));
      } else {
        formatted = options.map((opt) => { return { name: opt.title, value: opt.key } })
      }
      return (
        <SelectSearch
          key={questionType}
          onChange={this.handleSearchChange}
          options={formatted}
          placeholder="Search for a question"
          search={true}
        />
      );
    }
  };

  render() {
    const { questionType } = this.state
    return (
      <div className="box">
        <h4 className="title">Add New Activity</h4>
        <p className="control">
          <label className="label">Name</label>
          <input
            className="input"
            onChange={this.handleStateChange.bind(null, 'name')}
            placeholder="Text input"
            type="text"
            value={this.state.name}
          />
        </p>
        <p className="control">
          <label className="label">Landing Page Content</label>
        </p>
        <TextEditor
          ContentState={ContentState}
          EditorState={EditorState}
          handleTextChange={this.handleLPChange}
          shouldCheckSpelling={true}
          text={this.state.landingPageHtml || ''}
        />
        <br />
        <p className="control">
          <label className="label">Flag</label>
          <span className="select">
            <select defaultValue={this.state.flag} onChange={this.handleSelect}>
              <option value="alpha">alpha</option>
              <option value="beta">beta</option>
              <option value="gamma">gamma</option>
              <option value="production">production</option>
              <option value="archived">archived</option>
            </select>
          </span>
        </p>
        <p className="control">
          <label className="label">Question Type</label>
          <span className="select">
            <select defaultValue={questionType} onChange={this.handleSelectQuestionType}>
              <option value="questions">Sentence Combining</option>
              <option value="sentenceFragments">Sentence Fragment</option>
              <option value="fillInBlank">Fill In the Blank</option>
              <option value="titleCards">Title Cards</option>
            </select>
          </span>
        </p>
        <div className="control">
          <label className="label">Currently Selected Questions -- {`Total: ${this.state.selectedQuestions.length}`}</label>
          {this.renderQuestionSelect()}
        </div>
        <label className="label">All Questions</label>
        {this.renderSearchBox()}
        <br />
        <ChooseModelContainer
          conceptsFeedback={this.props.conceptsFeedback}
          modelConceptUID={this.state.modelConceptUID}
          updateModelConcept={modelConceptUID => this.setState({ modelConceptUID })}
        />
        <p className="control">
          <button className={`button is-primary ${this.props.stateSpecificClass}`} onClick={this.submit}>Submit</button>
        </p>
      </div>
    );
  }
}

function select(state) {
  return {
    questions: state.questions,
    concepts: state.concepts,
    sentenceFragments: state.sentenceFragments,
    conceptsFeedback: state.conceptsFeedback,
    fillInBlank: state.fillInBlank,
    titleCards: state.titleCards
  };
}

export default connect(select)(LessonForm);
