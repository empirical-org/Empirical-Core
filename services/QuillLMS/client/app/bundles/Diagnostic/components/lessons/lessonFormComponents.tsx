import * as React from 'react';

export interface DeleteButtonProps {
  questionId: string,
  onChange(questionId: string): void
}

export interface NameInputProps {
  name: string,
  onChange(value: string, e: object): void
}

export const DeleteButton = ({ questionId, onChange }: DeleteButtonProps) => {
  function handleClick() { onChange(questionId) }
  return <button onClick={handleClick} type="button">Delete</button>;
}

export const NameInput = ({ name, onChange }: NameInputProps) => {
  function handleChange(e: object) { onChange('name', e) };
  return(
    <label className="label" htmlFor="activity-name-input">
      Name
      <input
        aria-label="activity-name-input"
        className="input"
        id="activity-name-input"
        onChange={handleChange}
        placeholder="Text input"
        type="text"
        value={name}
      />
    </label>
  );
}
