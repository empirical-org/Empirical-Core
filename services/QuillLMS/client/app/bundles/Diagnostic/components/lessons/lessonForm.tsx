import { ContentState, EditorState } from 'draft-js';
import * as React from 'react';
import { connect } from 'react-redux';
import SelectSearch from 'react-select-search';
import {
  SortableList,
  TextEditor,
  hashToCollection
} from '../../../Shared/index';
import { Lesson } from '../../interfaces/lesson';
import { ConceptsReducerState } from '../../reducers/concepts';
import { ConceptsFeedbackReducerState } from '../../reducers/conceptsFeedback';
import { FillInBlankReducerState } from '../../reducers/fillInBlank';
import { QuestionsReducerState } from '../../reducers/questions';
import { SentenceFragmentsReducerState } from '../../reducers/sentenceFragments';
import { TitleCardsReducerState } from '../../reducers/titleCards';
import ChooseModel from './chooseModel';
import { DeleteButton, NameInput } from './lessonFormComponents.tsx';

export interface LessonFormProps {
  concepts: ConceptsReducerState,
  conceptsFeedback: ConceptsFeedbackReducerState,
  currentValues: Lesson,
  dispatch(action: any): any,
  fillInBlank: FillInBlankReducerState,
  lesson: Lesson,
  questions: QuestionsReducerState,
  sentenceFragments: SentenceFragmentsReducerState,
  stateSpecificClass?: string
  submit(object: {
    name: string,
    questions: Array<Object>,
    landingPageHtml: string,
    flag: string,
    modelConceptUID: string,
    isELL: boolean
  }): void,
  titleCards: TitleCardsReducerState,
}

export interface LessonFormState {
  flag: string,
  introURL: string,
  isELL: boolean,
  landingPageHtml: string,
  modelConceptUID: string,
  name: string,
  questionType: string,
  selectedQuestions: {
    key: string,
    questionType: string
  }[]
}


export class LessonForm extends React.Component<LessonFormProps, LessonFormState> {
  constructor(props) {
    super(props)

    const { flag, introURL, isELL, landingPageHtml, modelConceptUID, name, questions } = props.currentValues || {}; // eslint-disable-line react/destructuring-assignment

    this.state = {
      flag: flag || 'alpha',
      introURL: introURL || '',
      isELL: isELL || false,
      landingPageHtml: landingPageHtml || '',
      modelConceptUID: modelConceptUID || null,
      name: name || '',
      questionType: 'questions',
      selectedQuestions: questions || [],
    }
  }

  handleSubmit = () => {
    const { name, selectedQuestions, landingPageHtml, flag, modelConceptUID, isELL } = this.state;
    const { submit } = this.props;
    submit({
      name,
      questions: selectedQuestions,
      landingPageHtml,
      flag,
      modelConceptUID,
      isELL
    });
  }

  handleStateChange = (key: string, event: React.ChangeEvent<HTMLInputElement>) => {
    const changes = {};
    changes[key] = event.currentTarget.value;
    this.setState(changes);
  }

  handleQuestionChange = (value: string) => {
    const { selectedQuestions, questionType } = this.state;
    let newSelectedQuestions: any;
    const changedQuestion = selectedQuestions.find(q => q.key === value);
    if (!changedQuestion) {
      newSelectedQuestions = selectedQuestions.concat([{ key: value, questionType: questionType, }]);
    } else {
      newSelectedQuestions = selectedQuestions.filter(question => question !== changedQuestion);
    }
    this.setState({ selectedQuestions: newSelectedQuestions, });
  }

  handleSearchChange = (value: string) => {
    this.handleQuestionChange(value);
  }

  sortCallback = (sortInfo: {
    data: {
      items: {
        key: string,
        props: {
          defaultValue: string
        }
      }[]
    }
  }) => {
    const newOrder = sortInfo.map(item => {
      return { key: item.key, questionType: item.props.defaultValue }
    });
    this.setState({ selectedQuestions: newOrder });
  }

  renderQuestionSelect = () => {
    const { selectedQuestions } = this.state;
    // select changes based on whether we are looking at 'questions' (should be refactored to sentenceCombining) or sentenceFragments
    if (selectedQuestions && selectedQuestions.length) {
      const questionsList = selectedQuestions.map((question: { key: string, questionType: string }) => {
        const questionObject = this.props[question.questionType].data[question.key]; // eslint-disable-line react/destructuring-assignment
        const prompt = questionObject ? questionObject.prompt : 'Question No Longer Exists';
        const promptOrTitle = questionObject && question.questionType === 'titleCards' ? questionObject.title : prompt
        return (
          <p className="sortable-list-item" defaultValue={question.questionType} id={question.key} key={question.key}>
            {promptOrTitle}
            {'\t\t'}
            <DeleteButton onChange={this.handleQuestionChange} questionId={question.key} />
          </p>
        );
      });
      return <SortableList data={questionsList} id="currently-selected-questions" key={selectedQuestions.length} sortCallback={this.sortCallback} />;
    } else {
      return <div id="currently-selected-questions">No questions</div>;
    }
  }

  renderSearchBox = () => {
    const { questionType } = this.state;
    const { data } = this.props.concepts; // eslint-disable-line react/destructuring-assignment
    // options changes based on whether we are looking at 'questions' (should be refactored to sentenceCombining) or sentenceFragments
    let options = hashToCollection(this.props[questionType].data); // eslint-disable-line react/destructuring-assignment
    const concepts = data[0];
    let formatted: {}[];
    if (options.length > 0) {
      if (questionType !== 'titleCards') {
        options = options.filter((option: { conceptID: string, flag: string }) => {
          const concept = concepts.find(concept => concept.uid === option.conceptID);
          if(concept && option.flag !== 'archived') {
            return concept;
          }
        });
        formatted = options.map((opt: { key: string, prompt: string } )=> {
          return { name: opt.prompt.replace(/(<([^>]+)>)/ig, '').replace(/&nbsp;/ig, ''), value: opt.key, };
        });
      } else {
        formatted = options.map((opt: { key: string, title: string }) => {
          return { name: opt.title, value: opt.key }
        });
      }
      return (
        <SelectSearch
          id="all-questions"
          key={questionType}
          onChange={this.handleSearchChange}
          options={formatted}
          placeholder="Search for a question"
          search={true}
        />
      );
    }
  }

  handleSelectFlag = (e: React.FocusEvent<HTMLSelectElement>) => {
    this.setState({ flag: e.currentTarget.value, });
  }

  handleSelectQuestionType = (e: React.FocusEvent<HTMLSelectElement>) => {
    this.setState({ questionType: e.currentTarget.value, });
  }

  onLandingPageChange = (e: string) => {
    this.setState({ landingPageHtml: e, });
  }

  handleELLChange = () => {
    const { isELL } = this.state;
    this.setState({ isELL: !isELL });
  }

  onUpdateModelConcept = (uid: string) => {
    this.setState({ modelConceptUID: uid });
  }

  render() {
    const { flag, isELL, landingPageHtml, modelConceptUID, name, selectedQuestions } = this.state;
    const { conceptsFeedback, stateSpecificClass } = this.props;
    return (
      <div className="box">
        <h4 className="title">Add New Activity</h4>
        <p className="control">
          <NameInput name={name} onChange={this.handleStateChange} />
        </p>
        <div className="control">
          <label className="label" htmlFor="landing-page-content">Landing Page Content</label>
          <TextEditor
            handleTextChange={this.onLandingPageChange}
            id="landing-page-content"
            text={landingPageHtml || ''}
          />
        </div>
        <br />
        <p className="control">
          <label className="label" htmlFor="flag">
            Flag
            <span className="select">
              <select defaultValue={flag} onBlur={this.handleSelectFlag}>
                <option value="alpha">alpha</option>
                <option value="beta">beta</option>
                <option value="gamma">gamma</option>
                <option value="production">production</option>
                <option value="archived">archived</option>
              </select>
            </span>
          </label>
        </p>
        <p className="control">
          <label className="label" htmlFor="ell-check-input">
            Is ELL?
            <span>
              <input
                aria-label="ell-check-input"
                checked={isELL}
                onChange={this.handleELLChange}
                type="checkbox"
              />
            </span>
          </label>
        </p>
        <p className="control">
          <label className="label" htmlFor="question-type">
            Question Type
            <span className="select">
              <select defaultValue="questions" id="question-type" onBlur={this.handleSelectQuestionType}>
                <option value="questions">Sentence Combining</option>
                <option value="sentenceFragments">Sentence Fragment</option>
                <option value="fillInBlank">Fill In the Blank</option>
                <option value="titleCards">Title Cards</option>
              </select>
            </span>
          </label>
        </p>
        <div className="control">
          <label className="label" htmlFor="currently-selected-questions">
            Currently Selected Questions -- {`Total: ${selectedQuestions.length}`}
            {this.renderQuestionSelect()}
          </label>
        </div>
        <label className="label" htmlFor="all-questions">
          All Questions
          {this.renderSearchBox()}
        </label>
        <br />
        <ChooseModel
          conceptsFeedback={conceptsFeedback}
          modelConceptUID={modelConceptUID}
          updateModelConcept={this.onUpdateModelConcept}
        />
        <p className="control">
          <button className={`button is-primary ${stateSpecificClass}`} onClick={this.handleSubmit} type="submit">Submit</button>
        </p>
      </div>
    );
  }
};

function select({ questions, concepts, sentenceFragments, conceptsFeedback, fillInBlank, titleCards }) {
  return {
    questions,
    concepts,
    sentenceFragments,
    conceptsFeedback,
    fillInBlank,
    titleCards
  };
}

export default connect(select)(LessonForm);
