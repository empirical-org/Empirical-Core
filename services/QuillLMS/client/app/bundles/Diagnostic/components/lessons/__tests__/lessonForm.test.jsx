import { ContentState, EditorState } from 'draft-js';
import { shallow } from 'enzyme';
import React from 'react';
import {
    SortableList,
    TextEditor
} from '../../../../Shared/index';
import ChooseModel from '../chooseModel.tsx';
import { LessonForm } from '../lessonForm';
import { NameInput } from '../lessonFormComponents.tsx';

describe('LessonForm component', () => {
  const mockQuestions = [
    { key: '-KOqLeXEvMjNuE6MGOop', questionType: 'sentenceFragments' },
    { key: '-KdCgy8wt_rQiYpOURdW', questionType: 'fillInBlank' },
    { key: '-KdChHgbE9377Jgzkoci', questionType: 'fillInBlank' }
  ];
  const mockQuestionData = {
    '-KOqLeXEvMjNuE6MGOop': {
      conceptID: 'b5JpyWZYjhb94LTXGmYy7g',
      flag: 'production',
      key: '-KOqLeXEvMjNuE6MGOop',
      prompt: 'How much cheese is too much cheese?',
      title: 'Question 1'
    },
    '-KdCgy8wt_rQiYpOURdW': {
      conceptID: 'b5JpyWZYjhb94LTXGmYy7g',
      flag: 'production',
      key: '-KdCgy8wt_rQiYpOURdW',
      prompt: 'What is the best way to cultivate mass?',
      title: 'Question 2'
    },
    '-KdChHgbE9377Jgzkoci': {
      conceptID: 'b5JpyWZYjhb94LTXGmYy7g',
      flag: 'production',
      key: '-KdChHgbE9377Jgzkoci',
      prompt: 'Do jobs grow on jobbies?',
      title: 'Question 3'
    }
  };
  const mockProps = {
    currentValues: {
      flag: 'beta',
      isELL: true,
      landingPageHtml: '<p>Test content</p>',
      modelConceptUID: 'QsC1lua0t41_J2em_c7kUA',
      name: 'Awesome Diagnostic',
      questions: mockQuestions
    },
    lesson: {
      flag: 'beta',
      isELL: true,
      landingPageHtml: '<p>Test content</p>',
      modelConceptUID: 'QsC1lua0t41_J2em_c7kUA',
      name: 'Awesome Diagnostic',
      questions: mockQuestions
    },
    submit: jest.fn(),
    questions: {
      hasreceiveddata: true,
      submittingnew: false,
      states: {},
      data: mockQuestionData
    },
    concepts: {
      hasreceiveddata: true,
      submittingnew: false,
      states: {},
      data: {
        0: [{
          id: 589,
          name: 'Past Perfect',
          parent_id: 572,
          uid: 'b5JpyWZYjhb94LTXGmYy7g',
          description: null,
          level: 0,
          displayName: 'Verbs | Perfect Tense | Past Perfect'
        }]
      }
    },
    sentenceFragments: {
      hasreceiveddata: true,
      submittingnew: false,
      states: {},
      data: mockQuestionData
    },
    conceptsFeedback: {
      hasreceiveddata: true,
      submittingnew: false,
      states: {},
      data: mockQuestionData
    },
    fillInBlank: {
      hasreceiveddata: true,
      submittingnew: false,
      states: {},
      data: mockQuestionData
    },
    titleCards: {
      hasreceiveddata: true,
      data: {}
    },
    dispatch: jest.fn()
  };
  const container = shallow(<LessonForm {...mockProps} />);
  it('renders NameInput, TextEditor and ChooseModel components, passing expected props', () => {
    const handleStateChange = container.instance().handleStateChange;
    const onLandingPageChange = container.instance().onLandingPageChange;
    const handleUpdateModelConcept = container.instance().handleUpdateModelConcept;
    const nameInput= container.find(NameInput);
    const textEditor = container.find(TextEditor);
    const chooseModel = container.find(ChooseModel);

    expect(nameInput.length).toEqual(1);
    expect(TextEditor.length).toEqual(1);
    expect(chooseModel.length).toEqual(1);
    expect(nameInput.props().name).toEqual('Awesome Diagnostic');
    expect(nameInput.props().onChange).toEqual(handleStateChange);
    expect(textEditor.props().ContentState).toEqual(ContentState);
    expect(textEditor.props().EditorState).toEqual(EditorState);
    expect(textEditor.props().handleTextChange).toEqual(onLandingPageChange);
    expect(textEditor.props().text).toEqual('<p>Test content</p>');
    expect(chooseModel.props().conceptsFeedback).toEqual(mockProps.conceptsFeedback);
    expect(chooseModel.props().modelConceptUID).toEqual('QsC1lua0t41_J2em_c7kUA');
    expect(chooseModel.props().onUpdateModelConcept).toEqual(handleUpdateModelConcept);
  });
  it('handleSubmit calls submit() prop function passing name, questions: selectedQuestions, landingPageHtml, flag, modelConceptUID, and isELL as arguments', () => {
    const { name, selectedQuestions, landingPageHtml, flag, modelConceptUID, isELL } = container.state();
    const argument = {
      name,
      questions: selectedQuestions,
      landingPageHtml,
      flag,
      modelConceptUID,
      isELL
    };
    container.instance().handleSubmit();
    expect(mockProps.submit).toHaveBeenCalledWith(argument);
  });
  it('handleStateChange updates specific piece of state, using key and event arguments to build object to pass to setState', () => {
    const event = {
      currentTarget: {
        value: 'Even Moar Awesome Diagnostic'
      }
    };
    container.instance().handleStateChange('name', event);
    expect(container.state().name).toEqual(event.currentTarget.value);
  });
  it('handleQuestionChange updates selectedQuestions piece of state if question is added or deleted', () => {
    container.instance().handleQuestionChange('-KdChHgbE9212Jgzkoci');
    expect(container.state().selectedQuestions.length).toEqual(4);
    expect(container.state().selectedQuestions[3]).toEqual({ key: '-KdChHgbE9212Jgzkoci', questionType: 'questions'});
    container.instance().handleQuestionChange('-KdChHgbE9212Jgzkoci');
    expect(container.state().selectedQuestions.length).toEqual(3);
  });
  it('handleSearchChange calls handleQuestionChange, passing value as an argument', () => {
    const handleQuestionChange = jest.spyOn(container.instance(), 'handleQuestionChange');
    const value = 'saideira';
    container.instance().handleSearchChange(value);
    expect(handleQuestionChange).toHaveBeenCalledWith(value);
  });
  it('sortCallback sets selectedQuestion piece of state to reordered array of questions', () => {
    const sortInfo = [
      {
        $$typeof: 'Symbol(react.element)',
        type: 'p',
        key: '-KdCgy8wt_rQiYpOURdW',
        ref: null,
        props: {
          className: 'sortable-list-item',
          defaultValue: 'fillInBlank',
          children: []
        }
      },
      {
        $$typeof: 'Symbol(react.element)',
        type: 'p',
        key: '-KOqLeXEvMjNuE6MGOop',
        ref: null,
        props: {
          className: 'sortable-list-item',
          defaultValue: 'sentenceFragments',
          children: []
        }
      },
      {
        $$typeof: 'Symbol(react.element)',
        type: 'p',
        key: '-KdChHgbE9377Jgzkoci',
        ref: null,
        props: {
          className: 'sortable-list-item',
          defaultValue: 'fillInBlank',
          children: []
        }
      },
    ]
    const newOrder = [
      { key: '-KdCgy8wt_rQiYpOURdW', questionType: 'fillInBlank' },
      { key: '-KOqLeXEvMjNuE6MGOop', questionType: 'sentenceFragments' },
      { key: '-KdChHgbE9377Jgzkoci', questionType: 'fillInBlank' }
    ];
    container.setState({ selectedQuestions: mockQuestions });
    container.instance().sortCallback(sortInfo);
    expect(container.state().selectedQuestions).toEqual(newOrder);
  });
  it('renderQuestionSelect renders a SortableList component of selectedQuestions', () => {
    const keys = ['-KdCgy8wt_rQiYpOURdW', '-KOqLeXEvMjNuE6MGOop', '-KdChHgbE9377Jgzkoci'];
    const sortableList = container.find(SortableList);
    const getQuestionKey = i => sortableList.props().data[i].key;
    container.instance().renderQuestionSelect();
    expect(sortableList.length).toEqual(1);
    expect(getQuestionKey(0)).toEqual(keys[0])
    expect(getQuestionKey(1)).toEqual(keys[1])
    expect(getQuestionKey(2)).toEqual(keys[2])
  });
  it('renderSearchBox renders a SelectSearch component with question options', () => {
    const names = ['How much cheese is too much cheese?', 'What is the best way to cultivate mass?', 'Do jobs grow on jobbies?'];
    const handleSearchChange = container.instance().handleSearchChange;
    const selectSearch = container.find('#all-questions');
    const getQuestionName = i => selectSearch.props().options[i].name;
    container.instance().renderSearchBox();
    expect(container).toMatchSnapshot()
    expect(selectSearch.length).toEqual(1);
    expect(selectSearch.props().onChange).toEqual(handleSearchChange);
    expect(getQuestionName(0)).toEqual(names[0]);
    expect(getQuestionName(1)).toEqual(names[1]);
    expect(getQuestionName(2)).toEqual(names[2]);
  });
  it('handleSelectFlag updates the flag piece of state', () => {
    const e = {
      currentTarget: {
        value: 'alpha'
      }
    };
    container.instance().handleSelectFlag(e);
    expect(container.state().flag).toEqual(e.currentTarget.value);
  });
  it('handleSelectQuestionType updates the questionType piece of state', () => {
    const e = {
      currentTarget: {
        value: 'fillInBlank'
      }
    };
    container.instance().handleSelectQuestionType(e);
    expect(container.state().questionType).toEqual(e.currentTarget.value);
  });
  it('onLandingPageChange updates the landingPageHtml piece of state', () => {
    const e = '<p>Updated Content</p>'
    container.instance().onLandingPageChange(e);
    expect(container.state().landingPageHtml).toEqual(e);
  });
  it('handleELLChange updates isELL piece of state', () => {
    container.instance().handleELLChange();
    expect(container.state().isELL).toEqual(false);
  });
  it('onUpdateModelConcept updates modelConceptUID piece of state', () => {
    const id = 'new-id';
    container.instance().onUpdateModelConcept(id);
    expect(container.state().modelConceptUID).toEqual(id);
  });
});
