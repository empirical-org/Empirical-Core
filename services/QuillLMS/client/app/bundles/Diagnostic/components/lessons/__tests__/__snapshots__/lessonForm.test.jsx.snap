// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LessonForm component renderSearchBox renders a QuestionSelector component with question options 1`] = `
<div
  className="box"
>
  <h4
    className="title"
  >
    Add New Activity
  </h4>
  <p
    className="control"
  >
    <NameInput
      name="Even Moar Awesome Diagnostic"
      onChange={[Function]}
    />
  </p>
  <div
    className="control"
  >
    <label
      className="label"
      htmlFor="landing-page-content"
    >
      Landing Page Content
      <TextEditor
        ContentState={[Function]}
        EditorState={[Function]}
        handleTextChange={[Function]}
        id="landing-page-content"
        text="<p>Test content</p>"
      />
    </label>
  </div>
  <br />
  <p
    className="control"
  >
    <label
      className="label"
      htmlFor="flag"
    >
      Flag
      <span
        className="select"
      >
        <select
          defaultValue="beta"
          onBlur={[Function]}
        >
          <option
            value="alpha"
          >
            alpha
          </option>
          <option
            value="beta"
          >
            beta
          </option>
          <option
            value="gamma"
          >
            gamma
          </option>
          <option
            value="production"
          >
            production
          </option>
          <option
            value="archived"
          >
            archived
          </option>
        </select>
      </span>
    </label>
  </p>
  <p
    className="control"
  >
    <label
      className="label"
      htmlFor="ell-check-input"
    >
      Is ELL?
      <span>
        <input
          aria-label="ell-check-input"
          checked={true}
          onChange={[Function]}
          type="checkbox"
        />
      </span>
    </label>
  </p>
  <p
    className="control"
  >
    <label
      className="label"
      htmlFor="question-type"
    >
      Question Type
      <span
        className="select"
      >
        <select
          defaultValue="questions"
          id="question-type"
          onBlur={[Function]}
        >
          <option
            value="questions"
          >
            Sentence Combining
          </option>
          <option
            value="sentenceFragments"
          >
            Sentence Fragment
          </option>
          <option
            value="fillInBlank"
          >
            Fill In the Blank
          </option>
          <option
            value="titleCards"
          >
            Title Cards
          </option>
        </select>
      </span>
    </label>
  </p>
  <div
    className="control"
  >
    <label
      className="label"
      htmlFor="currently-selected-questions"
    >
      Currently Selected Questions -- 
      Total: 3
      <SortableList
        data={
          Array [
            <p
              className="sortable-list-item"
              defaultValue="fillInBlank"
            >
              What is the best way to cultivate mass?
              		
              <DeleteButton
                onChange={
                  [MockFunction] {
                    "calls": Array [
                      Array [
                        "saideira",
                      ],
                    ],
                    "results": Array [
                      Object {
                        "type": "return",
                        "value": undefined,
                      },
                    ],
                  }
                }
                questionId="-KdCgy8wt_rQiYpOURdW"
              />
            </p>,
            <p
              className="sortable-list-item"
              defaultValue="sentenceFragments"
            >
              How much cheese is too much cheese?
              		
              <DeleteButton
                onChange={
                  [MockFunction] {
                    "calls": Array [
                      Array [
                        "saideira",
                      ],
                    ],
                    "results": Array [
                      Object {
                        "type": "return",
                        "value": undefined,
                      },
                    ],
                  }
                }
                questionId="-KOqLeXEvMjNuE6MGOop"
              />
            </p>,
            <p
              className="sortable-list-item"
              defaultValue="fillInBlank"
            >
              Do jobs grow on jobbies?
              		
              <DeleteButton
                onChange={
                  [MockFunction] {
                    "calls": Array [
                      Array [
                        "saideira",
                      ],
                    ],
                    "results": Array [
                      Object {
                        "type": "return",
                        "value": undefined,
                      },
                    ],
                  }
                }
                questionId="-KdChHgbE9377Jgzkoci"
              />
            </p>,
          ]
        }
        id="currently-selected-questions"
        key="3"
        sortCallback={[Function]}
      />
    </label>
  </div>
  <label
    className="label"
    htmlFor="all-questions"
  >
    All Questions
    <Memo(ForwardRef)
      id="all-questions"
      key="questions"
      onChange={[Function]}
      options={
        Array [
          Object {
            "name": "How much cheese is too much cheese?",
            "value": "-KOqLeXEvMjNuE6MGOop",
          },
          Object {
            "name": "What is the best way to cultivate mass?",
            "value": "-KdCgy8wt_rQiYpOURdW",
          },
          Object {
            "name": "Do jobs grow on jobbies?",
            "value": "-KdChHgbE9377Jgzkoci",
          },
        ]
      }
      placeholder="Search for a question"
    />
  </label>
  <br />
  <ChooseModel
    conceptsFeedback={
      Object {
        "data": Object {
          "-KOqLeXEvMjNuE6MGOop": Object {
            "conceptID": "b5JpyWZYjhb94LTXGmYy7g",
            "flag": "production",
            "key": "-KOqLeXEvMjNuE6MGOop",
            "prompt": "How much cheese is too much cheese?",
            "title": "Question 1",
          },
          "-KdCgy8wt_rQiYpOURdW": Object {
            "conceptID": "b5JpyWZYjhb94LTXGmYy7g",
            "flag": "production",
            "key": "-KdCgy8wt_rQiYpOURdW",
            "prompt": "What is the best way to cultivate mass?",
            "title": "Question 2",
          },
          "-KdChHgbE9377Jgzkoci": Object {
            "conceptID": "b5JpyWZYjhb94LTXGmYy7g",
            "flag": "production",
            "key": "-KdChHgbE9377Jgzkoci",
            "prompt": "Do jobs grow on jobbies?",
            "title": "Question 3",
          },
        },
        "hasreceiveddata": true,
        "states": Object {},
        "submittingnew": false,
      }
    }
    modelConceptUID="QsC1lua0t41_J2em_c7kUA"
    updateModelConcept={[Function]}
  />
  <p
    className="control"
  >
    <button
      className="button is-primary undefined"
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </p>
</div>
`;
